form-flow-valtimo/src/main/kotlin/com/ritense/valtimo/formflow/FormFlowProcessLinkTaskProvider.kt

```diff
@@ -16,25 +16,107 @@
 
     override fun supports(formLink: FormLink?): Boolean {
         return formLink is BpmnElementFormFlowIdLink
     }
 
     override fun getTaskResult(task: Task, formLink: FormLink): TaskOpenResult<FormFlowTaskOpenResultProperties> {
         val instances = formFlowService.findInstances(mapOf("taskInstanceId" to task.id))
-        assert(instances.size == 1)
-        return TaskOpenResult(FORM_FLOW_TASK_TYPE_KEY, FormFlowTaskOpenResultProperties(instances[0].id.id))
+        val instance = when (instances.size) {
+            0 -> createFormFlowInstance(task)
+            1 -> instances[0]
+            else -> throw IllegalStateException("Multiple form flow instances linked to task: ${task.id}")
+        }
+        return TaskOpenResult(FORM_FLOW_TASK_TYPE_KEY, FormFlowTaskOpenResultProperties(instance.id.id))
     }
-}
\ No newline at end of file
+
+    private fun createFormFlowInstance(task: Task): FormFlowInstance {
+        val additionalProperties: MutableMap<String, Any> = getAdditionalProperties(task)
+        val formFlowId = getFormAssociationByTask(task).formLink.formFlowId
+        val formFlowDefinition: FormFlowDefinition = getFormFlowDefinition(formFlowId)
+        return formFlowService.save(formFlowDefinition.createInstance(additionalProperties))
+    }
+
+    private fun getFormAssociationByTask(task: Task): FormAssociation {
+        val processDefinition = repositoryService.createProcessDefinitionQuery()
+            .processDefinitionId(task.processDefinitionId)
+            .singleResult()
+
+        val formAssociation = formAssociationService.getFormAssociationByFormLinkId(
+            processDefinition.key,
+            task.taskDefinitionKey
+        ).orElseThrow { IllegalStateException("No form association found. Process: '${processDefinition.key}', task: '${task.taskDefinitionKey}'") }
+
+        if (formAssociation.formLink !is BpmnElementFormFlowIdLink) {
+            throw IllegalStateException("Found form association is not of type 'BpmnElementFormFlowIdLink'. Type: ${formAssociation.formLink.javaClass.simpleName}, process: '${processDefinition.key}', task: '${task.taskDefinitionKey}'")
+        }
+
+        return formAssociation
+    }
+
+    private fun getFormFlowDefinition(formFlowId: String): FormFlowDefinition {
+        val formFlowIdAsArray = formFlowId.split(":")
+        if (formFlowIdAsArray.size != 2) {
+            throw IllegalArgumentException("Invalid Format found for formFlowId '${Strings.join(formFlowIdAsArray, ':' )}'. Form flow id must have format key:version")
+        }
+        return if (formFlowIdAsArray[1] == "latest") {
+            formFlowService.findLatestDefinitionByKey(formFlowIdAsArray[0])!!
+        } else {
+            formFlowService.findDefinition(
+                FormFlowDefinitionId(formFlowIdAsArray[0], formFlowIdAsArray[1].toLong())
+            )
+        }
+    }
+
+    private fun getAdditionalProperties(task: Task): MutableMap<String, Any> {
```

Comment:
Please also change the return type to Map if we're to change the createInstance method